lib/spack/spack/compiler.py

```diff
@@ -174,6 +176,103 @@ def check(exe):
+    _LIBPATH_ARG = re.compile(r'^[-/](LIBPATH|libpath):')
+
+    def _determine_implicit_link_paths(self):
+        try:
+            compilers = [self.cc, self.cxx, self.f77, self.fc]
+            first_compiler = filter(None, compilers)[0]
+        except IndexError:
+            # No compilers available.
+            return []
+
+        try:
+            tmpdir = tempfile.mkdtemp(prefix='spack-implicit-link-info')
+            fout = os.path.join(tmpdir, 'output')
+            fin = os.path.join(tmpdir, 'main.c')
+
+            with open(fin, 'w+') as csource:
+                csource.write(
+                    'int main(int argc, char* argv[]) { '
+                    '(void)argc; (void)argv; return 0; }\n')
+                csource.flush()
+            output = Executable(first_compiler)(self.verbose_flag, fin, '-o',
+                                                fout, output=str, error=str)
+
+            lib_search_paths = False
+            raw_link_dirs = []
+            tty.debug('parsing implicit link info')
+            for line in output.splitlines():
+                if lib_search_paths:
+                    if line.startswith('\t'):
+                        raw_link_dirs.append(line[1:])
+                        continue
+                    else:
+                        lib_search_paths = False
+                elif line.startswith('Library search paths:'):
+                    lib_search_paths = True
+
+                if not self._LINKER_LINE.match(line):
+                    pass
+                if self._LINKER_LINE_IGNORE.match(line):
+                    pass
+                tty.debug('linker line: %s' % line)
+
+                next_arg = False
+                for arg in line.split():
+                    if arg in ('-L', '-Y'):
+                        next_arg = True
+                        continue
+
+                    if next_arg:
+                        raw_link_dirs.append(arg)
+                        next_arg = False
+                        continue
+
+                    if self._LINK_DIR_ARG.match(arg):
+                        tty.debug('linkdir: %s' % arg)
+                        raw_link_dirs.append(arg[2:])
+                    elif self._LIBPATH_ARG.match(arg):
+                        tty.debug('libpath: %s', arg)
+                        raw_link_dirs.append(arg[9:])
+            tty.debug('found raw link dirs: %s' % ', '.join(raw_link_dirs))
+
+            implicit_link_dirs = set()
+            for link_dir in raw_link_dirs:
+                implicit_link_dirs.add(os.path.abspath(link_dir))
+            implicit_link_dirs = list(implicit_link_dirs)
+
+            tty.debug('found link dirs: %s' % ', '.join(implicit_link_dirs))
+            return implicit_link_dirs
+        finally:
+            shutil.rmtree(tmpdir, ignore_errors=True)
+
+    __implicit_link_paths = None
+
+    @property
+    def implicit_link_paths(self):
+        if self.__implicit_link_paths is not None:
```

Comment:
Why two underscores?  Is that intentional (ala https://shahriar.svbtle.com/underscores-in-python?). Seems like just one would be sufficient.

Reply:
I meant it to be private. Subclasses shouldn't care about the caching here; if they really want to they should override the `implicit_link_paths` property instead.